app/Http/Controllers/Admin/ModulesController.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace App\Http\Controllers\Admin;
+
+use App\Contracts\Controller;
+use App\Models\Module;
+use App\Services\ModuleService;
+use Illuminate\Filesystem\Filesystem;
+use Illuminate\Http\Request;
+use Illuminate\Support\Str;
+use Laracasts\Flash\Flash;
+use Nwidart\Modules\Json;
+
+class ModulesController extends Controller
+{
+    private $moduleSvc;
+
+    public function __construct(ModuleService $moduleSvc)
+    {
+        $this->moduleSvc = $moduleSvc;
+    }
+
+    /**
+     * Display a listing of the resource.
+     *
+     * @return mixed
+     */
+    public function index()
+    {
+        $modules = Module::all()->sortByDesc('id');
+        $new_modules = $this->scan();
+        return view('admin.modules.index', [
+            'modules'     => $modules,
+            'new_modules' => $new_modules,
+        ]);
+    }
+
+    public function create()
+    {
+        return view('admin.modules.create');
+    }
+
+    public function store(Request $request)
+    {
+        $array = [
+            'file'    => $request->file('module_file'),
+            'enabled' => $request->has('enabled'),
+        ];
+
+        $store = $this->moduleSvc->createModule($array);
+
+        if ($store == true) {
+            Flash::success('Module Installed Successfully!');
+        }
+        Flash::error('Something Went Wrong! Please check the structure again or the module already exists!');
+        return redirect(route('admin.modules.index'));
+    }
+
+    public function edit($id)
+    {
+        $module = (new Module())->find($id);
+        return view('admin.modules.edit', [
+            'module' => $module,
+        ]);
+    }
+
+    public function update($id, Request $request)
+    {
+        $this->moduleSvc->updateModule($id, $request->has('enabled'));
+        Flash::success('Module Status Changed!');
+        return redirect(route('admin.modules.index'));
+    }
+
+    public function enable(Request $request)
+    {
+        $this->moduleSvc->enableModule($request->input('name'));
+        Flash::success('Module Enabled!');
+        return redirect(route('admin.modules.index'));
+    }
+
+    public function destroy($id, Request $request)
+    {
+        $delete = $this->moduleSvc->deleteModule($id, $request->all());
+        if ($delete == true) {
+            Flash::success('Module Deleted Successfully!');
+            return redirect(route('admin.modules.index'));
+        }
+        Flash::error('Verification Failed!');
+        return redirect(route('admin.modules.edit', $id));
+    }
+
+    /**
+     * Get & scan all modules.
+     *
+     * @return array
+     */
+    public function scan()
```

Comment:
This should stay in the service, this type of logic doesn't belong in the controller. The controllers should be relatively thin (there is older code right now in phpVMS which has a ton of logic in the controller that needs to be moved). But this belongs in the service and I'll probably break it apart more in the future so it can be tested

Reply:
Oh, I see.
Got something new to learn.
Thanks!

Reply:
You're doin great, hopefully this is a learning experience on how to organize a huge feature like this!

Reply:
Thanks!